* TOC
{:toc}

## BIOINF3010/7150: Genomics Applications
{:.no_toc}

Semester 1 2020

### Practical Timetable

| **Week** | **Monday** | **Practical**                                     |
|----------|------------|---------------------------------------------------|
| **1**    | 2/3        | [Introduction to Bash] (Dan)                      |
| **2**    | 9/3        | [Read Quality Control] (Nathan) [SAMTools and alignments] (Jimmy) |
| **3**    | 16/3       | [SARS-CoV-2 Resequencing] (Nathan) [SARS-CoV-2 Short Read Assembly] (Nathan)         |
| **4**    | 23/3       | [Short and long read alignment] (Nathan) [E. coli K-12 Hybrid Genome Assembly] (Nathan)         |
| **5**    | 30/3       | [Bacterial genome assembly] (Lloyd)               |
| **6**    | 6/4        | [HiC analysis] (Lloyd/Ning)                       |
| **-**    |            |                                                   |
| **7**    | 27/4       | [Genome graphs] (Yassine)                         |
| **8**    | 4/5        | [BLAST analysis and databases] (Dave)             |
| **9**    | 11/5       | [Clinical genomics] (Jimmy)                       |
| **10**   | 18/5       | [Agricultural genomics] (?)                       |
| **11**   | 25/5       | [Population genetics] (Bastien)                   |
| **12**   | 1/6        | [Metagenomics 16S profiling] (Raphael)            |

[Introduction to Bash]: Practicals/Bash_Practicals/1_IntroBash.md
[Read Quality Control]: Practicals/Read_QC/read-qc.md
[SAMTools and alignments]: Practicals/Alignments_Practicals/alignment-cram.md
[SARS-CoV-2 Resequencing]: Practicals/resequencing/resequencing.md
[SARS-CoV-2 Short Read Assembly]: Practicals/short_read_assembly/short-read-assembly.md
[Short and long read alignment]: Practicals/short_long_alignment/short_long_alignment.md
[E. coli K-12 Hybrid Genome Assembly]: Practicals/hybrid_genome_assembly/index.md
[Bacterial genome assembly]: Practicals/
[HiC analysis]: Practicals/HiC/hicprac.md
[Genome graphs]: Practicals/
[BLAST analysis and databases]: Practicals/
[Clinical genomics]: Practicals/
[Agricultural genomics]: Practicals/
[Population genetics]: Practicals/
[Metagenomics 16S profiling]: Practicals/

## Assessment

### Assessment Tasks

| **Assessment**                                            | **Subject**         |
|-----------------------------------------------------------|---------------------|
| [Assessment 0](Assignments/Assignment_0/bash_questions.md)| Bash                |
| [Assessment 1](Assignments/Assignment_1/genome_sequencing.md)                                          | Genome sequencing     |
| [Assessment 2]()                                          | Experimental design |
| [Assessment 3]()                                          | Publishing a genome |
| [Assessment 4]()                                          | Ancient DNA         |
| [Assessment 5]()                                          | Metagenomics        |
| [Project]() (PG only)                                     | Complete Dataset    |

### Project

In this course, the following next-generation sequencing (NGS) datasets/protocols will be examined in detail:

- Whole genome sequencing/Resequencing
- SNV variant/structural variation analysis
- Enrichment/Capture sequencing (Methyl-capture, ChIPseq, RIPseq)
- Metagenomics/Microbial profiling

Each of these NGS approaches uses similar programs and analysis approaches, such as quality control (quality and sequencing adapter trimming), genome alignment, and downstream visualisation and statistical methods. For the project, you will take a published (or otherwise obtained) dataset and complete all the analysis tasks (from raw data to final results) during the course.

You have the freedom to choose any dataset you would like. You will prepare a final report that will be due at the end of the semester. The report should be prepared using RStudio as an Rmd document including all code needed to perform the analysis, and will include the standard components of a scientific report: introduction (background on the study and identification of the research hypothesis), methods (analysis steps and programs used), results (what you found) and discussion (how the results relate to the research hypothesis and the published literature). The Rmd document and a compiled knitted html will form the submission; marks will be awarded to the code and Rmd that you use.

| Section | Mark |
|---------|-----:|
| Abstract | 5% |
| Introduction and hypothesis |	10% |
| Methods | 20% |
| Results and Discussion | 30% |
| References | 5% |
| Analysis scripts | 30% |


### Assessment Checklist

Have you:

- [ ] Answered all the questions?
- [ ] Followed naming conventions for Assessments?
- [ ] Checked that you have not breached the [Academic Honesty Policy](http://www.adelaide.edu.au/policies/230/).
- [ ] Identified the work as yours?
	- Emails should have the course and assessment task names.
	- Documents should be named with your name, the course name and the assessment task.
	- Printed documents should have you name and the course and assessment task in the text/footer/header.
- [ ] Used appropriate electronic communication with assessors?
	- Emails should have a meaningful subject.
- [ ] Handed in the assignment before the due time (see MyUni)?

## Useful Links

[How To Ask Questions The Smart Way](http://www.catb.org/esr/faqs/smart-questions.html)

[How to write a good bug report](https://musescore.org/en/developers-handbook/how-write-good-bug-report-step-step-instructions)
